packages/contracts/contracts/L1/ValidatorManager.sol

```diff
@@ -0,0 +1,687 @@
+    mapping(address => uint128) internal _jail;
+
+    /**
+     * @notice A mapping of output index challenged successfully to pending challenge rewards.
+     */
+    mapping(uint256 => uint128) internal _pendingChallengeReward;
+
+    /**
+     * @notice A modifier that only allows L2OutputOracle contract to call.
+     */
+    modifier onlyL2OutputOracle() {
+        require(
+            msg.sender == address(L2_ORACLE),
+            "ValidatorManager: Only L2OutputOracle can call this function"
+        );
+        _;
+    }
+
+    /**
+     * @notice Modifier to check if the caller is the Colosseum contract.
+     */
+    modifier onlyColosseum() {
+        require(
+            msg.sender == L2_ORACLE.COLOSSEUM(),
+            "ValidatorManager: Only Colosseum can call this function"
+        );
+        _;
+    }
+
+    /**
+     * @notice Semantic version.
+     * @custom:semver 1.0.0
+     */
+    string public constant version = "1.0.0";
+
+    /**
+     * @notice Constructs the ValidatorManager contract.
+     *
+     * @param _constructorParams The constructor parameters.
+     */
+    constructor(ConstructorParams memory _constructorParams) {
+        require(
+            _constructorParams._minRegisterAmount <= _constructorParams._minStartAmount,
+            "ValidatorManager: min register amount should not exceed min start amount"
+        );
+
+        L2_ORACLE = _constructorParams._l2Oracle;
+        ASSET_MANAGER = _constructorParams._assetManager;
+        TRUSTED_VALIDATOR = _constructorParams._trustedValidator;
+        MIN_REGISTER_AMOUNT = _constructorParams._minRegisterAmount;
+        MIN_START_AMOUNT = _constructorParams._minStartAmount;
+        COMMISSION_RATE_MIN_CHANGE_SECONDS = _constructorParams._commissionRateMinChangeSeconds;
+        // Note that this value MUST be (SUBMISSION_INTERVAL * L2_BLOCK_TIME) / 2.
+        ROUND_DURATION_SECONDS = _constructorParams._roundDurationSeconds;
+        JAIL_PERIOD_SECONDS = _constructorParams._jailPeriodSeconds;
+        JAIL_THRESHOLD = _constructorParams._jailThreshold;
+        MAX_OUTPUT_FINALIZATIONS = _constructorParams._maxOutputFinalizations;
+        BASE_REWARD = _constructorParams._baseReward;
+    }
+
+    /**
+     * @inheritdoc IValidatorManager
+     */
+    function registerValidator(
+        uint128 assets,
+        uint8 commissionRate,
+        uint8 commissionMaxChangeRate
+    ) external {
+        require(
+            getStatus(msg.sender) == ValidatorStatus.NONE,
+            "ValidatorManager: already initiated validator"
+        );
+
+        require(
+            assets >= MIN_REGISTER_AMOUNT,
+            "ValidatorManager: need to register with at least min register amount"
+        );
+
+        require(
+            commissionRate <= COMMISSION_RATE_DENOM,
+            "ValidatorManager: the max value of commission rate has been exceeded"
+        );
+
+        require(
+            commissionMaxChangeRate <= COMMISSION_RATE_DENOM,
```

Comment:
I think this should be changed as follows
```suggestion
            commissionMaxChangeRate + commissionRate <= COMMISSION_RATE_DENOM,
```

Reply:
If a validator sets `commissionRate` to 3 and `commissionMaxChangeRate` to 100, he can change the `commissionRate` to 0 and change it to 100 next time. Here the `commissionRate` means only the initial value, so I don't think it should be added to the condition.

Reply:
I got what you mean, the suggested one would be quite unnecessary.